src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs

```diff
@@ -108,6 +111,17 @@ public static void BeginInvokeDotNet(JSRuntime jsRuntime, DotNetInvocationInfo i
             // Returned a task - we need to continue that task and then report an exception
             // or return the value.
             task.ContinueWith(t => EndInvokeDotNetAfterTask(t, jsRuntime, invocationInfo), TaskScheduler.Current);
+
+        }
+        else if(syncResult is ValueTask valueTaskResult)
+        {
+            valueTaskResult.AsTask().ContinueWith(t => EndInvokeDotNetAfterTask(t, jsRuntime, invocationInfo), TaskScheduler.Current);
+        }
+        else if (syncResult?.GetType().IsGenericType ?? false && syncResult?.GetType().GetGenericTypeDefinition() == typeof(ValueTask<>))
+        {
+            var innerTask = GetTaskByType(syncResult.GetType().GenericTypeArguments[0], syncResult);
```

Comment:
Could reduce some repeated evaluations. Also I think the logic is complex enough to warrant a comment to clarify the intent.

```suggestion
        else if (syncResult?.GetType() is { IsGenericType: true } syncResultType
            && syncResultType.GetGenericTypeDefinition() == typeof(ValueTask<>))
        {
            // It's a ValueTask<T>. We'll coerce it to a Task so that we can attach a continuation.
            var innerTask = GetTaskByType(syncResultType.GenericTypeArguments[0], syncResult);
```

Reply:
I agree this is much clearer